app/src/main/java/org/wikipedia/analytics/eventplatform/EventPlatformClient.java

```diff
@@ -0,0 +1,446 @@
+        event.setSessionId(AssociationController.getSessionId());
+        event.setAppInstallId(getAppInstallId());
+    }
+
+    /**
+     * OutputBuffer: buffers events in a queue prior to transmission
+     *
+     * Transmissions are not sent at a uniform offset but are shaped into
+     * 'bursts' using a combination of queue size and debounce time.
+     *
+     * These concentrate requests (and hence, theoretically, radio awake state)
+     * so as not to contribute to battery drain.
+     */
+    static class OutputBuffer {
+
+        private static List<Event> QUEUE = new ArrayList<>();
+
+        /*
+         * When an item is added to QUEUE, wait this many ms before sending.
+         *
+         * If another item is added to QUEUE during this time, reset the
+         * countdown.
+         */
+        private static final int WAIT_MS = 30000;
+
+        /*
+         * When QUEUE.size() exceeds this value TIMER becomes non-interruptable.
+         */
+        private static final int MAX_QUEUE_SIZE = 128;
+
+        /*
+         * IMPLEMENTATION NOTE: Java Timer will provide the desired asynchronous
+         * countdown after a new item is added to QUEUE.
+         */
+        private static Timer TIMER = new Timer();
+
+        /*
+         * IMPLEMENTATION NOTE: Java abstract TimerTask class requires a run()
+         * method be defined.
+         *
+         * The run() method is called when the Timer expires.
+         */
+        private static class Task extends TimerTask {
+            public void run() {
+                sendAllScheduled();
+            }
+        }
+
+        /**
+         * If sending is enabled, dequeue and call send() on all scheduled items.
+         */
+        static void sendAllScheduled() {
+            TIMER.cancel();
+
+            if (ENABLED) {
+                /*
+                 * All items on QUEUE are permanently removed.
+                 */
+                for (Event event : QUEUE) {
+                    send(event);
+                }
+                QUEUE = new ArrayList<>();
+            }
+        }
+
+        /**
+         * Schedule a request to be sent.
+         *
+         * @param event event data
+         */
+        static void schedule(Event event) {
+            /*
+             * Item is enqueued whether or not sending is enabled.
+             */
+            QUEUE.add(event);
+
+            if (ENABLED) {
+                if (QUEUE.size() >= MAX_QUEUE_SIZE) {
+                    /*
+                     * >= because while sending is disabled, any number of items
+                     * could be added to QUEUE without it emptying.
+                     */
+                    sendAllScheduled();
+                } else {
+                    /*
+                     * The arrival of a new item interrupts the timer and resets
+                     * the countdown.
+                     */
+                    TIMER.cancel();
+                    TIMER = new Timer();
```

Comment:
Does the timer need to be re-created here?

Reply:
I see no other way to restart it. I will add a purge just to clear the queue, but no alternate to re-creating.